feat: add context-menu

janhassel opened the pull request:
Closes #6830

#### Changelog

**New**

- Adds the following react components:
  - `ContextMenu`
  - `ContextMenuOption`
  - `ContextMenuDivider`
  - `SelectableContextMenuOption`
  - `ContextMenuRadioGroup`

janhassel committed:
chore(ContextMenu): scaffold new component

janhassel committed:
feat(context-menu): build basic structure and styling

janhassel committed:
feat(context-menu): build basic nesting support

janhassel committed:
feat(context-menu): use button markup

janhassel committed:
feat(context-menu): add support for disabled options

janhassel committed:
feat(context-menu): add support fort shortcuts

janhassel committed:
feat(context-menu): add support for icons

janhassel committed:
feat(context-menu): add divider component

janhassel committed:
feat(context-menu): add support for selectable option

janhassel committed:
fix(context-menu): also indent options when sibling is selectable

janhassel committed:
feat(context-menu): add support for radio groups

janhassel committed:
fix(context-menu): add support for initialSelectedItem in radio groups

janhassel committed:
fix(context-menu): also indent options when sibling is radio

janhassel committed:
feat(context-menu): add support for keyboard navigation

janhassel committed:
feat(context-menu): dynamically reverse menu direction

janhassel committed:
feat(context-menu): add props.onClose

janhassel committed:
chore(context-menu): add inline notification to story

janhassel committed:
fix(context-menu): fix styling issues

janhassel committed:
Merge branch 'master' into context-menu

janhassel committed:
refactor(context-menu): rename SelectableContextMenuOption to ContextMenuSelectableOption

janhassel committed:
fix(context-menu): don't overwrite checkbox or radio role

janhassel committed:
fix(context-menu): automatically close menu after item was clicked

janhassel committed:
fix(context-menu): improve behaviour with screenreaders

janhassel committed:
fix(context-menu): add onChange to radio group and checkbox

janhassel committed:
test(context-menu): update public-api and index snapshots

janhassel committed:
chore(context-menu): correct formatting

janhassel committed:
test(context-menu): add render tests

janhassel committed:
chore: run prettier

emyarod commented on packages/react/src/components/ContextMenu/ContextMenu.js:
```diff
@@ -0,0 +1,244 @@
+        onKeyDown={handleKeyDown}
+        onClick={handleClick}
+        data-level={level}
+        style={isRootMenu ? { left: `${x}px`, top: `${y}px` } : null}
```
these positioning styles are always taking the `x` and `y` prop values, we probably want to save new `x` and `y` values after making calculations in `willFit` so that the menu reverses correctly

![image](https://user-images.githubusercontent.com/8265238/101688776-2183e580-3a32-11eb-97fb-7f92739f0cbe.png)

![image](https://user-images.githubusercontent.com/8265238/101688801-28125d00-3a32-11eb-8fbb-85886c81ee4a.png)

emyarod reviewed (commented):
hi @janhassel thank you for contributing this component! have a few general UX questions about the component behavior as well as a few notes about the component interactions:

* Since the first menu item in the story is a radio group, the group automatically expands. Should this be the expected behavior?
* Similarly, when a radio group is focused during keyboard navigation, should it open automatically?
* When using the arrow keys to navigate the menu, should the selection wrap by default?
* What is the maximum level of nesting allowed for context menus? (general spec/UX question)
* Clicking on a parent node currently collapses the menu, but we probably do not want that behavior by default
* If the menu is triggered while the browsing area is not in focus, the focus ring does not appear on the first valid node

andreancardona reviewed (changes requested):
Hey Jan @janhassel - in reviewing this with @emyarod we noticed right clicking does not trigger the context-menu in iOS - let me know if I'm missing something.
![130549163_404558264317138_5500473654424346359_n](https://user-images.githubusercontent.com/32720851/101689086-8fc8a800-3a32-11eb-91d3-0db218b08c49.jpg)

carmacleod commented:
@janhassel 
#### For keyboard testing:

Please add a few focusable elements to the test page (for example, a button, link, and multi-line textarea), and show how the context menu can be correctly positioned over the focusable element when the user types <kbd>Shift + F10</kbd> (or <kbd>Context Menu</kbd> key if available) on Windows:
- centered horizontally/vertically over the button/link (seems to be the most common/intuitive location for non-editable-text elements)
- beside the caret in an editable text field/text area/contenteditable

Note that macOS doesn't have a way to open a context menu using only the keyboard (unless you are using the VoiceOver screen reader). (Although I heard a rumor that Edge for Mac might support <kbd>Shift + F10</kbd> - I may try that out. If it's true, I might even switch to Edge as my default browser on Mac! 😄 ).

----

To try to answer some of @emyarod's questions:
> Since the first menu item in the story is a radio group, the group automatically expands. Should this be the expected behavior?

No. Submenus should not automatically expand when opened. They can be opened when mouse hovers over the trigger (after a "hover intent" timeout). Perhaps that is what is happening here (i.e. maybe mouse hover location is contained within the context menu's bounds), however note that other context menu implementations do not show any hover effects on open (so either they are waiting for the user to move the mouse, or they are positioning the menu a pixel or two away from the mouse location on open - I don't know which).

Regarding when to expand if using keyboard (e.g. <kbd>Down arrow</kbd> or <kbd>Up arrow</kbd> to select menuitem that has submenu), the submenu should not be automatically expanded - instead, for keyboard, it should be expanded & first item focused on <kbd>Right arrow</kbd>, and closed on <kbd>Left arrow</kbd> & return focus to menuitem that triggered submenu.

> Similarly, when a radio group is focused during keyboard navigation, should it open automatically?

No. Context menus are not opened on focus. (Not sure why specifically radio group? Did I misunderstand the question?)

> When using the arrow keys to navigate the menu, should the selection wrap by default?

Yes. Selection should wrap in context menu, same as in all menus.

> What is the maximum level of nesting allowed for context menus? (general spec/UX question)

One. See [Mac Guidelines](https://developer.apple.com/design/human-interface-guidelines/ios/controls/context-menus/): "Keep submenus to one level.".
The [Windows win32 UX Guidelines](https://docs.microsoft.com/en-us/windows/win32/uxguide/cmd-menus#context-menus) say "Avoid using submenus to keep context menus simple, direct, and efficient." and the example immediately above this text shows a "Size" submenu. :)  However, I think we can safely say that both sets of guidelines strongly discourage more than one level of nesting in a context menu.

> Clicking on a parent node currently collapses the menu, but we probably do not want that behavior by default

Actually, I think context menus always close when you click away from them? At least, they do on Windows. Will try on Mac in a few minutes.

> If the menu is triggered while the browsing area is not in focus, the focus ring does not appear on the first valid node

I don't see this on Windows, but I will test on Mac next.

Regarding @andreancardona's question:
> right clicking does not trigger the context-menu in iOS

Wow, are you [using a mouse with your iOS device](https://support.apple.com/en-ca/HT210546)? I've never tried that. :)
I've only ever used a "touch & hold" gesture on the touch screen to [open a context menu on iOS](https://developer.apple.com/design/human-interface-guidelines/ios/controls/context-menus/). Apparently 3D touch can also open a context menu on newer iOS devices (mine is old).
Do you also have a bluetooth keyboard connected to your iOS device? If so, try <kbd>Control</kbd> + left click (i.e. half keyboard, half mouse). I am only guessing that that might work on iOS, because <kbd>Control</kbd> + left click can be used on macOS as an alternative to right click (because Mac mouse used to have only one button).

carmacleod commented:
#### Keyboard testing on Mac:
Installed Edge for Mac, and was delighted to note that it _does_ support <kbd>Shift + F10</kbd> to open context menu! 🎉 
Of course, this shortcut is not supported in Safari, Chrome, or Firefox on Mac. :(

On Mac, the context menu in this PR opens with either right-click or <kbd>Control</kbd> + left-click in Chrome, Edge and Firefox, but only right-click works in Safari. When I try to use <kbd>Control</kbd> + left-click to open the example context menu in Safari, the menu opens but then it closes immediately. This must be a bug somewhere in the PR code, because <kbd>Control</kbd> + left-click has no trouble opening a context menu anywhere on this github page.

Confirmed that context menus on Mac close when you click anywhere outside the menu, just like context menus on Windows.

> If the menu is triggered while the browsing area is not in focus, the focus ring does not appear on the first valid node

I do see this on Mac. However, none of the other context menu implementations that I tested show any hover or focus effects when the menu is triggered. The first <kbd>Down arrow</kbd> after opening selects (focuses) the first non-disabled menu item. So I don't think this context menu should have a focus ring (or selection or hover highlight) on any node right after the menu is triggered. Wait for the user to either move mouse or type arrow keys before highlighting/focusing menu items.

janhassel commented:
Thank you, @emyarod @andreancardona @carmacleod for your thoughts!

On a side note: I think you can enable a global right-click shortcut in the macOS accessibility settings: ![image](https://user-images.githubusercontent.com/28265588/101761880-faceb900-3adc-11eb-891e-584ef74359b8.png)

In general I think we need to align on "how far" this component goes because I think we had different assumptions. I purposely built this component without the right-click trigger logic built in and just demoed it that way in the storybook as it would be the recommended use. I wanted to keep it up to the team though how to trigger this component as I can think of some use cases (at least in our application) where a simple `window.addEventListener('contextmenu')` might not be sufficient. I could render the context menu by default in the storybook to make this more clear.

Points I'll try to tackle in the next few days:
- Making sure it is always rendered on the visible screen area
- Not auto-focusing the first option but the menu itself and waiting for up or down arrow
- Only show submenu on right-arrow, not on focus
- Incorporating a hover-intent delay for opening submenus

I also think that we'd need a completely different design for SM breakpoint which maybe more closely resembles the overlaying actions sliding up from the bottom on iOS or the vertically centered actions on Android.

emyarod commented:
> Not auto-focusing the first option but the menu itself and waiting for up or down arrow

I think auto focusing the first option is fine, but if the item is a nested menu then it shouldn't open by default. would that be a better user experience?

carmacleod commented:
@janhassel Thanks for planning to tackle those items!

Good question about whether the component includes the trigger logic or not. Would be nice if the component defaulted to triggering on [contextmenu event](https://developer.mozilla.org/en-US/docs/Web/API/Element/contextmenu_event), though, so that devs wouldn't have to reinvent the most common trigger logic everywhere (particularly the bit about positioning the context menu based on the focused element if invoked with keyboard). Maybe some prop that lets you provide your own logic (which would turn off default triggering). I'm not on the Carbon team, though, so I'm just guessing here. @emyarod, can you cc the right folks to discuss trigger logic API?

Jan, can you explain the alternate triggering use case that you mentioned?

Also worth discussing is that there seem to be several different ways to handle contextmenu event on the web. Since the browser already provides a context menu, we are essentially adding an additional "app" context menu on top of the browser's. The user may - or may not - want to access the browser's context menu (maybe to open a link in a new window, maybe to inspect an element in devtools, maybe to fix spelling, etc). This choice may be app specific, or even specific to the particular context menu. So Carbon may have to provide a prop for what to do with the browser's context menu. I have seen:
- simply preventDefault and don't open the browser's context menu at all. This is what google docs and codesandbox do, and some of the context menus in Slack web. (And I see that this is what the current Storybook example does).
- open the Carbon context menu on the first right click (or shift+F10), and then open the browser's context menu on the 2nd right click (or shift+f10) if the mouse cursor (or focus) hasn't moved. This is what Box does, and some of the other context menus in Slack web.
- I think I saw a third option somewhere a while ago, but I can't find it now. It was an app context menu where the last menu item ("More items...") opened the browser's context menu. Maybe I'm dreaming. ;)

Jan, regarding the macOS pointer control settings, thank-you. I've seen that before, but I couldn't figure out how to change right-click to be invoked on Shift+F10 instead of F12. I use F12 to open devtools all the time - would have to change my brain when using Mac. 😄  However, now that I look at it again, I might try F12 for a while (maybe I can get used to <kbd>Option+Command+I</kbd> for devtools).

@emyarod 
> I think auto focusing the first option is fine, but if the item is a nested menu then it shouldn't open by default. would that be a better user experience?

None of the other context menus that I tested focus the first option on open (including web pages like Box, Google docs, Slack web, codesandbox, and also including browser/native desktop context menus on both Windows and Mac). We probably shouldn't either, for consistency.

janhassel committed:
fix(context-menu): focus entire menu on open

janhassel committed:
fix(context-menu): only open submenu on long hover or right arrow

janhassel committed:
feat(context-menu): always keep menu within body boundaries

janhassel commented:
I implemented the points I mentioned in my last comment.

Additionally, I changed the fixed-width of 13rem to a min-width and set the max-width to 20rem (320px) so that a longer label would stretch the context menu wihout being immediately truncated. This comes from concerns around translation (For example: Something simple like "Cut" would be "Ausschneiden" in German and would basically already fill the entire width of the menu if paired with an icon).

@carmacleod
> Jan, can you explain the alternate triggering use case that you mentioned?

What I'm mostly thinking about is pairing this with third-party libraries like ag-grid (for advanced data grids) or codemirorr (for code editors). Usually, these libraries already provide a context menu but allow the developer to replace it with their own component. In that case the library still handles the `contextmenu` event and may pass additional information (which cell is selected, what code is highlighted, ...).

Regarding the trigger logic: I don't think this should be a part of the component itself due to the concerns I mentioned above as well as the developer experience it would bring (devs needing to make sure there is always only one ContextMenu component mounted to avoid all of them being called by the event for example). However, the longer I think about it, the more it seems to be a good candidate for a react hook. I think the Carbon team wanted to explore hook opportunities like this in the future. I could imagine something like the following (@joshblack I'd love your input on this):
```js
import React, { useRef } from 'react';
import { useContextMenu } from '@carbon/react';

function SomeComponent() {
  const trigger = useRef(null);
  const contextMenuProps = useContextMenu(trigger);

  return (
    <> 
      <div ref={trigger}>Hello world</div>
      <ContextMenu {...contextMenuProps}>
        ...
      </Contextmenu>
    </>
  );
}

export default SomeComponent;
```
This way we could have a standardized trigger logic with options like @carmacleod mentioned (for example our application should never show the browser context menu as we're running on the desktop) while keeping it separate from the component for better developer experience and more flexibility with third-party libaries like the ones mentioned above.

laurenmrice reviewed (changes requested):
This is fantastic, @janhassel !! 🤩 🙏 Overall it looks great. Just a couple of visual tweaks that need fixing. 

---------

- The caret on the right needs to come down 1px or 2px to align inline with the text. Right now it is positioned a little too high.
![caret misaligned](https://user-images.githubusercontent.com/43969356/104774677-b9b17080-573c-11eb-9306-3d1143383025.png)

- Focused items in the menu should not have a hover background color under the focus border (unless this was done for accessibility purposes for this kind of component specifically?).
![focus with no background hover](https://user-images.githubusercontent.com/43969356/104774717-c930b980-573c-11eb-8379-0e01dd8e4509.png)

- The entire menu gets a thin blue focus outline. Does this have to be visible as a color to know that it is in focus?

- On hover: disabled options in the menu should not get a visible hover background, only a disabled cursor change.
![disabled item no hover color](https://user-images.githubusercontent.com/43969356/104774815-efeef000-573c-11eb-8a30-8a74ff04d32d.png)

- What kind of command icons are you using here? Are these the standard icons that should be used for this? Just wondering because I am not entirely sure.
![command icons](https://user-images.githubusercontent.com/43969356/104774851-fd0bdf00-573c-11eb-8f4e-79309eb8aabe.png)

- In Safari, when I `control and click`, the menu does not stay open. 

- Can we change the "Publish" icon from a checkmark to a 16px [report icon](https://carbon-elements.netlify.app/icons/examples/preview/#Report32) for demo purposes. Just to not confuse people that the item can be checked or unchecked.
![publish icon](https://user-images.githubusercontent.com/43969356/104775184-8d4a2400-573d-11eb-85ff-e740302347c1.png)

- The context menu you are showing here is the one that can contain icons before the text in each menu item. I think the default version we show should be the context menu without icons before the text. And then we can show this example as a different variation. Maybe show this as a prop to turn on or as another option in the same dropdown menu under the knobs tab? @emyarod 
![show the default version without icons](https://user-images.githubusercontent.com/43969356/104774925-1c0a7100-573d-11eb-9526-a388a6712cb3.png)

- Question about max-width:
> Additionally, I changed the fixed-width of 13rem to a min-width and set the max-width to 20rem (320px) so that a longer label would stretch the context menu wihout being immediately truncated. This comes from concerns around translation (For example: Something simple like "Cut" would be "Ausschneiden" in German and would basically already fill the entire width of the menu if paired with an icon).

Would changing the max-width to 18rem (288px) instead of 20rem (320px) be acceptable? Or is that still too small.

janhassel committed:
fix(context-menu): implement review feedback

janhassel committed:
fix(context-menu): don't close when disabled or divider was clicked

janhassel commented:
@laurenmrice Thanks for the review! See updates below.

> The caret on the right needs to come down 1px or 2px to align inline with the text. Right now it is positioned a little too high.

✅  Fixed

> Focused items in the menu should not have a hover background color under the focus border (unless this was done for accessibility purposes for this kind of component specifically?).

✅  Fixed

> The entire menu gets a thin blue focus outline. Does this have to be visible as a color to know that it is in focus?

💬 Exactly, this follows the behaviour of list boxes (such as [https://react.carbondesignsystem.com/?path=/story/dropdown--default](Dropdown)). The reason is that once the entire menu is in focus, the user can either press <kbd>arrow down</kbd> or <kbd>arrow up</kbd> resulting in either the first or last item being focused.

> On hover: disabled options in the menu should not get a visible hover background, only a disabled cursor change.

✅  Fixed

> What kind of command icons are you using here? Are these the standard icons that should be used for this? Just wondering because I am not entirely sure.

💬 These are the unicode symbols macOS is using for their shortcuts, meaning they're normal text: ⌥ ⌘ ⌫ ⎋ ⇧ ⇪

> In Safari, when I control and click, the menu does not stay open.

✅  Fixed

> Can we change the "Publish" icon from a checkmark to a 16px report icon for demo purposes. Just to not confuse people that the item can be checked or unchecked.

💬 Actually... it is a checkbox / can be checked an unchecked. These types of "inline" checkbox options are quite common for context menus. I agree that is hard to spot, maybe we'll need to drop icon support so that only checkmarks are allowed (controlled by the component)?

![image](https://user-images.githubusercontent.com/28265588/104935984-36965180-59ac-11eb-993f-c26cce061510.png)

> The context menu you are showing here is the one that can contain icons...

✅  Made the default story contain no icons and created a second story "with icons"

> Question about max-width: Would changing the max-width to 18rem (288px) instead of 20rem (320px) be acceptable? Or is that still too small.

✅  Fixed. In case teams run into issues with translation, they can still overwrite the `max-width` but then need to make sure it won't grow wider than the screen.

janhassel committed:
fix(context-menu): prevent menu from closing immediately in Safari

laurenmrice reviewed (changes requested):
> The entire menu gets a thin blue focus outline. Does this have to be visible as a color to know that it is in focus?

> 💬 Exactly, this follows the behaviour of list boxes (such as [https://react.carbondesignsystem.com/?path=/story/dropdown--default](Dropdown)). The reason is that once the entire menu is in focus, the user can either press <kbd>arrow down</kbd> or <kbd>arrow up</kbd> resulting in either the first or last item being focused.

✅ Ok great !


> What kind of command icons are you using here? Are these the standard icons that should be used for this? Just wondering because I am not entirely sure.

> 💬 These are the unicode symbols macOS is using for their shortcuts, meaning they're normal text: ⌥ ⌘ ⌫ ⎋ ⇧ ⇪

✅  Perfect


> Can we change the "Publish" icon from a checkmark to a 16px report icon for demo purposes. Just to not confuse people that the item can be checked or unchecked.

> I agree that is hard to spot, maybe we'll need to drop icon support so that only checkmarks are allowed (controlled by the component)?

Okay, I see what you mean now. Yes, I agree with what you said above ^.

------------------------------------------------------------------------------------------

- [x] The menu still autocloses in Safari for me when I control click.

![Jan-21-2021 14-23-09](https://user-images.githubusercontent.com/43969356/105408181-3dafa080-5bf4-11eb-8e54-4bbc6498c43a.gif)

janhassel committed:
refactor(context-menu): separate between external and internal component

janhassel committed:
refactor(context-menu): mark as unstable

janhassel committed:
feat(context-menu): export useContextMenu hook

janhassel committed:
test(context-menu): replace Option component with Item component

janhassel committed:
Merge branch 'master' into context-menu

janhassel commented:
@laurenmrice I updated the component and story to only allow icons from checkboxes and radio groups.

> The menu still autocloses in Safari for me when I control click.

Seems to be happening when a click is longer than x milliseconds. With very short right-clicks it works. This is still due to the Safari emitting the `click` event in addition to the `contextmenu` event when `preventDefault()` is called within the `contextmenu` event. I honestly don't have a real solution for this one yet. Maybe I'll need to wait for `mouseup` before allowing the menu to be closed by a click. I'll see if can work on that in the coming week.

@joshblack I exported the `useContextMenu` hook and all components with the `unstable_` prefix as discussed. The hook should ideally support more options (see @carmacleod's comment above: https://github.com/carbon-design-system/carbon/pull/7350#issuecomment-742838011). I think this can be handled in separate PRs in the future though before releasing the component as stable. What do you think?
Also, let me know if there are any conventions around experimental components that I missed!

janhassel committed:
test: update react index snapshot

laurenmrice commented:
Thanks @janhassel ! Can you also add in the option to view the version of the context menu `with icons` ? I think it disappeared.

janhassel commented:
@laurenmrice I thought we agreed on dropping icon support and keep it exclusive to the checkmark for single-select and multi-select options.

> > I agree that is hard to spot, maybe we'll need to drop icon support so that only checkmarks are allowed (controlled by the component)?

> Okay, I see what you mean now. Yes, I agree with what you said above ^.

Let me know if I misunderstood something.

laurenmrice commented:
@janhassel Oh okay, yes I did misunderstand what you said about removing ALL icon support. 😅 Sorry, Jan. I was thinking if someones design requires having a `checkmark` selection in the main menu, then we should specify that then they should not be using the variant of the context menu that has other icons next to the text. Maybe this is just more for written guidance and does not need to be specified in the code? 

***Edit: Spoke with Jan via slack and we have decided to not have icon support for this version. We can think about adding in icon support as an enhancement at a later date once this component has been released for awhile and we have some more concrete guidance around when to use icons. 👍🏻

laurenmrice reviewed (approved):
Looks so great! Thank you Jan for contributing this 🎉 ⭐️

emyarod reviewed (commented):
one question about interaction: when a user presses <kbd>Enter</kbd> on a radio group, should the group then open?

janhassel committed:
fix(context-menu): open submenus on enter

janhassel commented:
@emyarod Yes, definitely! Good point! Just pushed an update

janhassel committed:
fix(context-menu): fix lint issue

emyarod reviewed (approved):
thanks for the updates, looks good to me

andreancardona committed:
Merge branch 'master' into context-menu

andreancardona reviewed (approved):
Thank you @janhassel  for such an awesome contribution ❤️ 🎉 !

carmacleod commented:
Accessibility review:
- please change all `role="checkbox"` to `role="menuitemcheckbox"`
- please change all `role="radio"` to `role="menuitemradio"`
- please delete all `aria-label`s on menuitems. They are unnecessary because a menuitem's label comes from its contents (and you do not need to override characters like ⌘ because screen readers all know what ⌘ is, and they say "command").
- please delete all `title` attributes. They are not necessary (and they are annoying 😄).

The Windows screen readers kinda-sorta work ok, but it is unexpected to hear "button" when focusing a menuitem, which is happening because the ARIA markup is a little bit "wonky". :)

With ARIA markup, all of the ARIA needs to go on the focusable element (i.e. the one with tabindex).
However, the menuitems in this context menu have 2 different elements sharing the ARIA markup and the tabindex:
an `li` and a `button`. 

There are 2 ways you could go to fix this. Either:
- put all of the menuitem markup and the tabindex on the `li`, and strip out the `button`. For an example of doing it this way, please see the menus in the [APG Editor menubar example](https://w3c.github.io/aria-practices/examples/menubar/menubar-editor.html).
- OR put `role="none"` on the `li`, and all of the menuitem markup and the tabindex on the `button`. For an example of doing it this way please see the menus in the [APG Navigation menubar example](https://w3c.github.io/aria-practices/examples/menubar/menubar-navigation.html) (except that you would be using button where they use link).

carmacleod reviewed (changes requested):
In the submenu, please delete the div with `role="radiogroup"`, and move its aria-label up to the div's parent `ul` to give the menu a name that matches it's triggering menuitem. The radiogroup role is not used in ARIA menus. The `group` role is allowed in menus, but it is only useful for separating a group of menuitems from other menuitems in the menu - i.e. similar function to a separator - so `role="group"` wouldn't do anything in this particular submenu because there's only 1 group.

Up/Down arrows wrap in Windows menus, but they don't wrap in Mac menus. I don't know what happens on other platforms. (Best guess is that it wraps on Linux platforms because they tend to have similar keyboard behavior to Windows). Would it be possible to special-case whether or not to wrap based on platform?

This is looking really nice @janhassel !

janhassel committed:
fix(context-menu): use menuitem* roles for checkbox and radio

janhassel committed:
fix(context-menu): remove aria-label from menuitems

janhassel committed:
refactor(context-menu): remove button markup

janhassel committed:
test(context-menu): update test to reflect markup changes

janhassel committed:
fix(context-menu): use ul without role for radio groups

janhassel commented:
@carmacleod Thanks for the extensive review! Really appreciate it. 🙏 

> please change all role="checkbox" to role="menuitemcheckbox"
> please change all role="radio" to role="menuitemradio"

💬  I used these roles in the beginning but reverted them back to `checkbox` and `radio` as `menuitemcheckbox` and `menuitemradio` are not properly communicated by VoiceOver, instead they both get read as "menu item".
I pushed an update, changing them back to to the `menuitem*` roles. Let me know what you think and if we should go with the spec or with what has better support right now (also let me know how JAWS reads it if you have access to a windows machine).

> please delete all aria-labels on menuitems. They are unnecessary because a menuitem's label comes from its contents (and you do not need to override characters like ⌘ because screen readers all know what ⌘ is, and they say "command").

✅  Done

> please delete all title attributes. They are not necessary (and they are annoying 😄).

💬  I agree that they can be annoying, but I have to disagree about their lack of importance here. We cannot assume that a label will always fit into the maximum width of this menu, especially if the label contains the name of an item to edit or when we talk about internationalization.

Here is a quick example from the Finder context menu
![image](https://user-images.githubusercontent.com/28265588/107757643-1d23c380-6d26-11eb-81f9-d1af41465dc9.png)
![image](https://user-images.githubusercontent.com/28265588/107757593-0e3d1100-6d26-11eb-8489-d05cc1a5373e.png)


And a context menu from the Messages.app on macOS (in German)
![image](https://user-images.githubusercontent.com/28265588/107756721-ec8f5a00-6d24-11eb-931d-bdc410db5d85.png)
![image](https://user-images.githubusercontent.com/28265588/107756697-e00b0180-6d24-11eb-84e5-0bc18f87a01e.png)

> put all of the menuitem markup and the tabindex on the li, and strip out the button

✅  Done

> In the submenu, please delete the div with role="radiogroup", and move its aria-label up to the div's parent ul

💬  I removed the role and changed the implementation from a `div` to a `ul`. The reason I cannot add the label to the parent node and remove the extra markup is that a radio group is not exclusive to a submenu. Like in the examples you showed, having multiple radio groups and other menu items on the same level is allowed. Let me know if there are any additional attributes needed on the `ul`.

![image](https://user-images.githubusercontent.com/28265588/107757969-886d9580-6d26-11eb-8ffd-c4fb7d787e99.png)
![image](https://user-images.githubusercontent.com/28265588/107758330-02058380-6d27-11eb-8b0d-808a7c69c6a8.png)

> Would it be possible to special-case whether or not to wrap based on platform?

💬  Good point! Do you know if native dropdowns also wrap on windows? Because dropdowns in Carbon do not wrap right now (tested on macOS only though) and I was following these existing "list of items" components for the interaction.

carmacleod commented:
Hi, thanks for the updates, @janhassel !
Sorry to take so long to get back to you.

I'm having trouble testing completely because the aria-haspopup on the "Share with" menuitem keeps getting set to `aria-haspopup="[object Object]"`... please change it to either `aria-haspopup="true"` or `aria-haspopup="menu"`. The Object value is confusing the screen readers into thinking the submenu is just a list (they don't recognize it as a menu).

Also, the "Share with" menuitem `li` needs to have `aria-expanded="true"` when the menu is open (and "false" when it is closed).

The submenu `ul` should have `aria-label="Share with"` OR it could have `aria-labelledby="share-with-menuitem-li-id"` (where you would need to add a unique `id="share-with-menuitem-li-id"` to the "Share with" menuitem `li`).

I think that once we get the structure exactly right, the menuitemcheckbox and menuitemradio roles will work on all platforms. [ARIA menu markup](https://w3c.github.io/aria-practices/#wai-aria-roles-states-and-properties-13) is extremely picky! 😄 

Regarding `title`, can you only add it when there has been truncation? I have never seen a context menu that had tooltips popping up for every menu item. (Actually, I have never seen a context menu that truncates - I thought they would grow to accommodate all menu items - but you're absolutely right that it's possible, particularly if there are filenames in the context menu.)

Regarding groups in menus:
- yes, multiple groups are allowed, but each one needs a `role="group" aria-label="Group name"`
- if all of the menuitems in a menu belong to the same group, as is the case for your "Share with" menu, then you would end up with "Share with" menu (because menus need a label), followed by "Share with" group, which is redundant. So in that case, it is easier to just put all of the menuitems into the menu `ul` and not have a separate group `ul`, to avoid having to duplicate the labels. This is what we see in the APG Editor Menubar example's Font menu (1st screen snapshot, below).
- if, however, there are multiple groups in the same menu, e.g. as in the Style/Color menu in the APG Editor Menubar example (2nd screen snapshot), then you can have a `ul role="group" aria-label="Group name"`, but note that they need to be inside an `<li role="none">` so that the HTML is valid (i.e. ul as direct child of ul is not valid html).

#### APG Editor Menubar example showing Font menu markup (no group needed)
![image](https://user-images.githubusercontent.com/3331913/108740874-1eae7200-7504-11eb-855f-fd8bdfb3d4ca.png)

#### APG Editor Menubar example showing Style/Color menu markup (with Font style group and Text color group)
![image](https://user-images.githubusercontent.com/3331913/108735737-e48ea180-74fe-11eb-9bef-7302a2e011b0.png)


> Do you know if native dropdowns also wrap on windows?

Yep, they sure do.

One more thing: regarding <kbd>Space</kbd> key vs. <kbd>Enter</kbd> key for activating a menu item. Typically, <kbd>Enter</kbd> is used for activating a control (such as a menu item) on Windows/Linux platforms, and <kbd>Space</kbd> is used for activation on Mac. I notice that your context menu activates menu items on <kbd>Enter</kbd>. You probably want to support <kbd>Space</kbd> as well, which is more familiar to Mac users (although I notice that Mac seems to support both <kbd>Space</kbd> and <kbd>Enter</kbd> activation for menu items, so probably the easiest thing is to just support both).

You could consider also having <kbd>Space</kbd> change the state of menuitemcheckbox and menuitemradio items _without closing the menu_ - but that could be something to consider for a follow-on issue to this one - definitely optional, and not required! (it's a feature of some menus)

janhassel committed:
fix(context-menu): fix aria-haspopup attribute on parent node

janhassel committed:
fix(context-menu): add aria-expanded attribute on items with children

janhassel committed:
fix(context-menu): add "group" role to radio groups

janhassel committed:
fix(context-menu): extract radio items from group when top level

janhassel committed:
chore: add story with multiple radio groups

janhassel committed:
feat(context-menu): add ContextMenuGroup component

janhassel committed:
fix(context-menu): support "space" to activate items

janhassel commented:
@carmacleod

> aria-haspopup on the "Share with" menuitem keeps getting set to aria-haspopup="[object Object]"

✅ Fixed

> the "Share with" menuitem li needs to have aria-expanded="true" when the menu is open (and "false" when it is closed).

✅ Done

> The submenu ul should have aria-label="Share with"

💬 Let me know if I misunderstand, but the `ul` has an `aria-label` on my side. Maybe this is fixed with the new markup for top-level groups (see end of list)
![image](https://user-images.githubusercontent.com/28265588/108822425-5b8b6f00-75bf-11eb-9990-09a1d62e1664.png)

> Regarding title, can you only add it when there has been truncation?

💬 I would really like to avoid this as it adds a good amount of extra logic to determine in JavaScript whether or not a truncation will happen (which in turn is done by CSS though). I think this can be a point we can bring up at a later stage. The component will still be marked as unstable/experimental for a bit.

> multiple groups are allowed, but each one needs a role="group" aria-label="Group name"

✅ Done

> if all of the menuitems in a menu belong to the same group...
> if, however, there are multiple groups in the same menu...

✅ Done

> > Do you know if native dropdowns also wrap on windows?

> Yep, they sure do.

💬 Good to know! Do Carbon dropdowns wrap on Windows as well? I think we should follow this consistently.

> You probably want to support Space as well, which is more familiar to Mac users

✅ Done

tw15egan committed:
Merge branch 'master' into context-menu

carmacleod commented:
@janhassel Thanks for the really fast fixes! Sorry to take so long to get back to you. This is looking really good now!

I did notice something odd in the example: After the context menu has been opened and closed, the user needs to right-click twice to pop up the menu again. Or Shift+F10 twice. Or ctrl+left click (on Mac) twice.

>> The submenu ul should have aria-label="Share with"

> Let me know if I misunderstand, but the ul has an aria-label on my side. Maybe this is fixed with the new markup for top-level groups

Yes, you fixed it with the new markup for top-level groups. :)

>> Regarding title, can you only add it when there has been truncation?

> 💬 I would really like to avoid this as it adds a good amount of extra logic ...

https://github.com/carbon-design-system/carbon/pull/7768 might already have some of that extra logic?
Let me know if you prefer that I open a new issue to look at this later.

>>> Do you know if native dropdowns also wrap on windows?

>> Yep, they sure do.

> 💬 Good to know! Do Carbon dropdowns wrap on Windows as well? I think we should follow this consistently.

Good question. It looks like Combobox does, and Dropdown doesn't. Do people use Dropdown to create menus  in Carbon?
I notice that [OverflowMenu](https://deploy-preview-7350--carbon-components-react.netlify.app/?path=/story/components-overflowmenu--basic) _does_ wrap. How should we proceed with this? New issue to discuss what to do across components?

> 💬 I used these roles in the beginning but reverted them back to checkbox and radio as menuitemcheckbox and menuitemradio are not properly communicated by VoiceOver, instead they both get read as "menu item".
I pushed an update, changing them back to to the menuitem* roles. Let me know what you think and if we should go with the spec or with what has better support right now (also let me know how JAWS reads it if you have access to a windows machine).

I think we have to go with the spec, because otherwise we will mess up the Windows screen readers. VO seems to have some strange quirks in this area, like [saying "1" when a radio or checkbox menuitem is checked](https://bugs.webkit.org/show_bug.cgi?id=222916). Those will get fixed in time. The [ARIA-AT project](https://github.com/w3c/aria-at/wiki/Background) is working towards more consistency of AT behavior across platforms.

Just for info, here's what NVDA says now (in Chrome) for the radio and checkbox menuitems. For the radio menu items, it says "radio menu item", and it says both "checked" and "not checked". Note that it doesn't say "checkbox" menuitem, but it does say both checked and not checked.
```
menu
Share with  subMenu  1 of 9
expanded
None  radio menu item  not checked  1 of 4
Share with  menu
Product team  radio menu item  checked  2 of 4
...
Publish  checked  7 of 9
space
...
Publish  not checked  7 of 9
```

Here's what JAWS says (in Chrome) for those items (ignore the weird whitespace characters). Note that JAWS doesn't say radio or checkbox. For checkboxes, it says both "checked" and "not checked", but for radios, it only says "checked" and it doesn't say "not checked".
```
menu 
To navigate press Up or Down Arrow.
Share with sub menu 1 of 9
None 1 of 4
Product team checked 2 of 4
...
Publish not checked 7 of 9
Leaving menus
...
Publish checked 7 of 9
```

lorelei-ngooi commented:
Is there any ETA on when this would be available?

janhassel committed:
Merge branch 'main' into context-menu

janhassel committed:
chore: add context menu story to experimental group

janhassel committed:
fix(context-menu): prevents error where user must click twice to open

janhassel commented:
@carmacleod 

> After the context menu has been opened and closed, the user needs to right-click twice to pop up the menu again

✅ Good catch, should be fixed now

>>> Regarding title, can you only add it when there has been truncation?

> Let me know if you prefer that I open a new issue to look at this later.

💬 Yes, please open an issue for that once this PR is merged in

> Good question. It looks like Combobox does, and Dropdown doesn't. Do people use Dropdown to create menus in Carbon?
I notice that OverflowMenu does wrap. How should we proceed with this? New issue to discuss what to do across components?

💬 Let's definitely open a serparate issue for this one. Personally, I think these should all behave the same, but a member of the Carbon team needs to make a call and then it should be applied to the individual components.

> Just for info, here's what NVDA says now [...]

💬 Thanks, this is really helpful! Let's hope VoiceOver, JAWS and co will adopt these newer roles soon.

---

@emyarod I believe this means the PR is ready for review again. Let's bring this in as experimental / unstable soon and work on further concerns and ideas in separate, smaller (and more managable) PRs.
Let me know if you have any questions or if you wanna schedule a quick chat!

emyarod reviewed (commented):
yes I have been following this after I approved the first time, just requested a review from myself while awaiting @carmacleod 's reviews. regarding the menu selection wrap behavior, I think the context menu should also wrap selection but that can be addressed separately

janhassel commented:
@emyarod Awesome. Yes, let's handle the wrapping in a separate issue.

carmacleod reviewed (approved):
+1 - good to go!

kodiakhq[bot] committed:
Merge branch 'main' into context-menu

carmacleod commented:
A bit of follow-up housekeeping:
- opened https://github.com/carbon-design-system/carbon/issues/8073 for the title issue
- opened https://github.com/carbon-design-system/carbon/issues/8076 for the wrap issue

janhassel commented:
Thank you, @carmacleod! 🙂
